File: C++/delayed-count-of-equal-elements.cpp
// Time:  O(n)
// Space: O(n)

// freq table
class Solution {
public:
    vector<int> delayedCount(vector<int>& nums, int k) {
        vector<int> result(size(nums));
        unordered_map<int, int> cnt;
        for (int i = size(nums) - 1 - k; i >= 0; --i) {
            result[i] = cnt[nums[i]];
            ++cnt[nums[i + k]];
        }
        return result;
    }
};

File: Python/delayed-count-of-equal-elements.py
# Time:  O(n)
# Space: O(n)

import collections


# freq table
class Solution(object):
    def delayedCount(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        result = [0]*len(nums)
        cnt = collections.defaultdict(int)
        for i in reversed(xrange(len(nums)-k)):
            result[i] = cnt[nums[i]]
            cnt[nums[i+k]] += 1
        return result
